### PR TITLE
Add possibility to use http2 preface with https connection

### LINKED ISSUE #4702: Option to force HTTP/2 connection over TLS without negotiation
Hello, is it possible to force the Armeria HTTP Client to use HTTP/2 over TLS and skip negotiation, or even if negotiation fails? I mean if I for sure know that the server support HTTP/2.
Is it possible to send a preface in such a case?
https://httpwg.org/specs/rfc7540.html#known-http

### PR DESCRIPTION
**Pull request by 0lejk4**
Motivation:

Add the possibility to use http2 preface with HTTPS connection for cases when the server does not support ALPN negotiation

Modifications:

- HttpClientPipelineConfigurator will try connecting with HTTP/2 Preface when the option is enabled and downgrade to HTTP/1 if the server does not support it.
- Changed HttpSession.retryWithH1C to HttpSession.retryWith method that receives with which SessionProtocol to retry
- Changed HttpSessionHandler to comply with HttpSession change

Result:

- Closes #4702 
- Users can use the HTTP/2 Preface option for secure connections

### CODE REVIEW

**Comment by trustin** [comment]
Hey, thanks a lot for your PR. I think it makes sense to add a way to deal with an HTTP/2 server that doesn't support ALPN. However, it'd be nice if this new behavior is enabled only when a user explicitly tells so. Currently, `useHttp2Preface` is `true` by default, which means the server that explicitly told the client that it doesn't support HTTP/2 will get HTTP/2 connection preface.

Therefore, could you add adding a new option to `ClientFactoryOptions` for this behavior? e.g. `ClientFactoryOptions.USE_HTTP2_WITHOUT_ALPN` (disabled by default)

I also would like to ask the following changes:

- There are two ways to start an HTTP/2 connection - 1) HTTP/2 connection preface and 2) HTTP/1-to-2 upgrade request. Your PR supports only the former. Could you also handle the case (2) when `useHttp2Preface` is `false`?
- Could you add a test case for the new option? It should test the following 4 scenarios:
  | | HTTP/2 server without ALPN | HTTP/1-only server without ALPN |
  |-|----------------------------|---------------------------------|
  | useHttp2Preface = true<br>(Connection preface) | ✔ | ✔* |
  | useHttp2Preface = false<br>(Upgrade request) | ✔ | ✔* |

  *: Also make sure `SessionProtocolNegotiationCache` is updated accordingly.

**Comment by 0lejk4** [comment]
Wanted to clarify, so I can reuse the Upgrade logic from:
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java#L313
And extract it to some upgrade method and same for this:
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java#L295
So the solution will look somewhat like this:
Screenshot 2023-02-27 at 10 16 10

**Comment by jrhee17** [comment]
> So the solution will look somewhat like this:

I think the code you shared looks similar to how I imagined it 👍

**Comment by 0lejk4** [comment]
Committed the USE_HTTP2_WITHOUT_ALPN option support and now is time to deal with test cases for it. Is there some similar test case that I could look into? If not I guess I need to start the server with tlsSelfSigned and mock io.netty.handler.ssl.SslHandler#applicationProtocol on the client to simulate negotiation failure. Any thoughts on this?

**Comment by ikhoon** [comment]
> Is there some similar test case that I could look into?

You can build a simple HTTP/2 over TLS server using the combination of `NettyServerExtension`,  `SelfSignedCertificateExtension` and `SimpleH2CServerHandler`. A `SniHandler` and `ApplicationProtocolNegotiationHandler` need to be prepended to `SimpleH2CServerHandler`.

For example:
```java
@RegisterExtension
static SelfSignedCertificateExtension ssc = new SelfSignedCertificateExtension();

@RegisterExtension
static NettyServerExtension h2Server = new NettyServerExtension() {
    @Override
    protected void configure(Channel ch) throws Exception {

        final SslContext sslContext = SslContextBuilder
                .forServer(ssc.privateKey(), ssc.certificate())
                .build();
        ch.pipeline().addLast(new SniHandler(new DomainWildcardMappingBuilder<>(sslContext).build()));
        ch.pipeline().addLast(new ApplicationProtocolNegotiationHandler("???") {

            @Override
            protected void configurePipeline(ChannelHandlerContext ctx, String protocol) throws Exception {
                ctx.pipeline().addLast(new H2ConnectionHandlerBuilder().build());
            }
        });
    }
};

class H2ConnectionHandler extends SimpleH2CServerHandler {
    H2ConnectionHandler(Http2ConnectionDecoder decoder, Http2ConnectionEncoder encoder,
                        Http2Settings initialSettings) {
        super(decoder, encoder, initialSettings);
    }
}

class H2ConnectionHandlerBuilder
        extends AbstractHttp2ConnectionHandlerBuilder<H2ConnectionHandler, H2ConnectionHandlerBuilder> {

    @Override
    protected H2ConnectionHandler build() {
        return super.build();
    }

    @Override
    protected H2ConnectionHandler build(Http2ConnectionDecoder decoder, Http2ConnectionEncoder encoder,
                                        Http2Settings initialSettings) throws Exception {
        final H2ConnectionHandler handler = new H2ConnectionHandler(decoder, encoder,
                                                                    initialSettings);
        frameListener(handler);
        return handler;
    }
}
}
```

**Comment by ikhoon** [comment]
> Do you have any ideas? Currently stuck on it(

A `channelActive` event makes `UpgradeRequestHandler` send `OPTIONS * HTTP/1.1`.
https://github.com/line/armeria/blob/f2da76b695592a74c766e629faea33ebe629401e/core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java#L426-L432
It is working for HTTP since `UpgradeRequestHandler` is added before a channel gets active.
However, `UpgradeRequestHandler` is added after the TLS handshake for HTTPS, so `UpgradeRequestHandler` can't get a `channelActive` event. As a workaround, we can use `handlerAdded()` event for HTTPS.

**Comment by trustin** [comment]
The server we're trying to test should actually not support ALPN, so it's probably better assuming that you write a plaintext HTTP/2 server that supports upgrade requests, and then insert an `SslHandler` in the beginning of Netty pipeline, with ALPN disabled. You probably might want to check how we implemented plaintext HTTP in `HttpServerPipelineConfigurator`..?

**Comment by 0lejk4** [comment]
> > Do you have any ideas? Currently stuck on it(
> 
> A `channelActive` event makes `UpgradeRequestHandler` send `OPTIONS * HTTP/1.1`.
> 
> https://github.com/line/armeria/blob/f2da76b695592a74c766e629faea33ebe629401e/core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java#L426-L432
> 
> 
> It is working for HTTP since `UpgradeRequestHandler` is added before a channel gets active.
> However, `UpgradeRequestHandler` is added after the TLS handshake for HTTPS, so `UpgradeRequestHandler` can't get a `channelActive` event. As a workaround, we can use `handlerAdded()` event for HTTPS.

Yes, this worked, thanks, but making working test was not an easy challenge for me) It looks weird but this works:

```java
class H2ConnectionHandler extends SimpleH2CServerHandler {
    protected final Http2Connection.PropertyKey streamKey;
    private final Http2Connection.PropertyKey upgradeKey;
    private final InboundHttpToHttp2Adapter adapter;

    H2ConnectionHandler(Http2ConnectionDecoder decoder, Http2ConnectionEncoder encoder,
                        Http2Settings initialSettings) {
        super(decoder, encoder, initialSettings);
        streamKey = connection().newKey();
        upgradeKey = connection().newKey();
        adapter = new InboundHttpToHttp2Adapter(connection(), this);
    }

    @Override
    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
        super.userEventTriggered(ctx, evt);
        if(evt instanceof UpgradeEvent) {
            UpgradeEvent upgrade = (UpgradeEvent) evt;
            try {
                ctx.fireUserEventTriggered(upgrade.retain());
                Http2Stream stream = connection().stream(HTTP_UPGRADE_STREAM_ID);
                upgrade.upgradeRequest().headers().setInt(HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), HTTP_UPGRADE_STREAM_ID);
                stream.setProperty(upgradeKey, true);
                adapter.channelRead(ctx, upgrade.upgradeRequest().retain());
            } finally {
                upgrade.release();
            }
        }
    }
}
```

Is there any simpler way to achieve this or this is good enough for this test case?

**Comment by ikhoon** [comment]
As Trustin said, a pipeline customer would be a simple solution to set up a test server. #4710
It would not take a long time to fix #4710. How about continuing your work when #4710 is handled? We are working on it ASAP.

**Comment by 0lejk4** [comment]
Understood! Good, then I`ll wait for it, thanks for the clarification.

**Comment by ikhoon** [comment]
Gentle ping @0lejk4 😉

**Comment by 0lejk4** [comment]
Sorry for inconvenience, I am really sorry I couldn't help you in time. I see that you are finishing this on your own. Still glad this feature is coming to life, thank you for your work guys!

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java [hunk 1]
```diff
@@ -727,6 +728,16 @@ public ClientFactoryBuilder useHttp2Preface(boolean useHttp2Preface) {
         return this;
     }
 
+    /**
+     * Sets whether to use HTTP/2 without ALPN. This is useful if you want to communicate with an HTTP/2
+     * server over TLS but the server does not support ALPN.
+     */
+    @UnstableApi
+    public ClientFactoryBuilder useHttp2WithoutALPN(boolean useHttp2WithoutALPN) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java:736]:
```suggestion
    public ClientFactoryBuilder useHttp2WithoutAlpn(boolean useHttp2WithoutAlpn) {
```

File: core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java [hunk 2]
```diff
@@ -491,6 +499,14 @@ public boolean useHttp2Preface() {
         return get(USE_HTTP2_PREFACE);
     }
 
+    /**
+     * Returns whether to use HTTP/2 over TLS without ALPN.
+     */
+    @UnstableApi
+    public boolean useHttp2WithoutALPN() {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java:506]:
```suggestion
    public boolean useHttp2WithoutAlpn() {
```

File: core/src/main/java/com/linecorp/armeria/client/HttpClientFactory.java [hunk 3]
```diff
@@ -101,6 +101,7 @@ final class HttpClientFactory implements ClientFactory {
     private final long maxConnectionAgeMillis;
     private final int maxNumRequestsPerConnection;
     private final boolean useHttp2Preface;
+    private final boolean useHttp2WithoutALPN;
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/HttpClientFactory.java:104]:
```suggestion
    private final boolean useHttp2WithoutAlpn;
```

File: core/src/main/java/com/linecorp/armeria/client/HttpClientFactory.java [hunk 4]
```diff
@@ -260,6 +262,10 @@ boolean useHttp2Preface() {
         return useHttp2Preface;
     }
 
+    boolean useHttp2WithoutALPN() {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/HttpClientFactory.java:265]:
```suggestion
    boolean useHttp2WithoutAlpn() {
```

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 5]
```diff
@@ -774,6 +777,22 @@ public static boolean defaultUseHttp2Preface() {
...
+     * This will be either HTTP/2 connection preface or HTTP/1-to-2 upgrade request,
+     * depending on {@link ClientFactoryBuilder#useHttp2Preface(boolean)} setting.
+     * If disabled, when ALPN negotiation fails client will also fail in case HTTP/2 was required.
+     * {@link ClientFactoryBuilder#useHttp2WithoutALPN(boolean)}.
+     *
+     * <p>This flag is disabled by default. Specify the
+     * {@code -Dcom.linecorp.armeria.defaultUseHttp2WithoutALPN=true} JVM option to enable it.
+     */
+    @UnstableApi
+    public static boolean defaultUseHttp2WithoutALPN() {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java:792]:
```suggestion
    public static boolean defaultUseHttp2WithoutAlpn() {
```

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 6]
```diff
@@ -774,6 +777,22 @@ public static boolean defaultUseHttp2Preface() {
...
+    /**
+     * Returns the default value of the {@link ClientFactoryBuilder#useHttp2WithoutALPN(boolean)} option.
+     * If enabled, even when ALPN negotiation fails client will try to attempt upgrade to HTTP/2 when needed.
+     * This will be either HTTP/2 connection preface or HTTP/1-to-2 upgrade request,
+     * depending on {@link ClientFactoryBuilder#useHttp2Preface(boolean)} setting.
+     * If disabled, when ALPN negotiation fails client will also fail in case HTTP/2 was required.
+     * {@link ClientFactoryBuilder#useHttp2WithoutALPN(boolean)}.
+     *
+     * <p>This flag is disabled by default. Specify the
+     * {@code -Dcom.linecorp.armeria.defaultUseHttp2WithoutALPN=true} JVM option to enable it.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java:789]:
```suggestion
     * {@code -Dcom.linecorp.armeria.defaultUseHttp2WithoutAlpn=true} JVM option to enable it.
```

**Review by trustin — COMMENTED** [review]
Left some naming convention comments, but it looks great. Thanks, @0lejk4 and @ikhoon! :bow:

**Review by trustin — APPROVED** [review]
:+1:

**Review by jrhee17 — APPROVED** [review]
Looks really nice 👍    Thanks @0lejk4 and @ikhoon ! 🙇 👍 🚀

**Review by ikhoon — APPROVED** [review]
Thanks, @0lejk4! 👍

**Review by minwoox — APPROVED** [review]
Thanks a lot! 👍

**Comment by minwoox** [comment]
Thanks, @0lejk4 for adding this feature. 🙇

**Pull request merged**